// prime/cov/defer.js
var __$coverCall, __$coverInit, __$coverInitRange;
(function(fs){
	"use strict";
	var update = function(cb){
		return function(a, b, c, d){
			var json = JSON.parse(fs.readFileSync("/home/arian/www/prime/cov.json").toString() || '{}');
			cb(json, a, b);
			fs.writeFileSync("/home/arian/www/prime/cov.json", JSON.stringify(json, null, 2));
		};
	};
	__$coverInit = update(function(json, name, code){
		if (!json[name]) json[name] = {__code: code};
	});
	__$coverInitRange = update(function(json, name, range){
		if (!json[name][range]) json[name][range] = 0;
	});
	__$coverCall = update(function(json, name, range){
		json[name][range]++;
	});
})(require('fs'));
__$coverInit("defer.js", "/*\ndefer\n*/\"use strict\"\n\nvar deferred = {\n    timeout: {},\n    frame: {},\n    immediate: {}\n}\n\nvar UID = 0\n\nvar _defer = function(type, callback, arg){\n\n    var collection = deferred[type],\n        method     = functions[type]\n\n    var empty = true\n    for (var p in collection){\n        empty = false;\n        break;\n    }\n    var uid = (UID++).toString(36)\n    if (empty) method(arg)\n    collection[uid] = callback\n    return function(){\n        delete collection[uid]\n    }\n\n}\n\nvar iterate = function(collection){\n    var time = new Date().getTime()\n\n    var exec = {}\n    for (var p in collection){\n        exec[p] = collection[p]\n        delete collection[p]\n    }\n    for (var p in exec) exec[p](time)\n}\n\nvar immediate = function(){\n    iterate(deferred.immediate)\n}\n\nvar functions = {}\n\nif (global.process && process.nextTick){\n    functions.immediate = function(){\n        process.nextTick(immediate)\n    }\n} else if (global.setImmediate){\n    functions.immediate = function(){\n        setImmediate(immediate)\n    }\n} else if (global.postMessage && global.addEventListener){\n\n    addEventListener(\"message\", function(event){\n        if (event.source === global && event.data === \"@deferred\"){\n            event.stopPropagation()\n            immediate()\n        }\n    }, true)\n\n    functions.immediate = function(){\n        postMessage(\"@deferred\", \"*\")\n    }\n\n} else {\n    functions.immediate = function(){\n        setTimeout(immediate, 0)\n    }\n}\n\nvar requestAnimationFrame = global.requestAnimationFrame ||\n    global.webkitRequestAnimationFrame ||\n    global.mozRequestAnimationFrame ||\n    global.oRequestAnimationFrame ||\n    global.msRequestAnimationFrame ||\n    function(callback) {\n        setTimeout(callback, 1e3 / 60)\n    }\n\nfunctions.frame = function(){\n    requestAnimationFrame(function(){\n        iterate(deferred.frame)\n    })\n}\n\nfunctions.timeout = function(ms){\n    var timeout = deferred.timeout\n    var collection = timeout[ms] || (timeout[ms] = {})\n    setTimeout(function(){\n       iterate(collection)\n    }, ms)\n}\n\n/*\nexport!\n*/\n\nvar defer = function(callback, ms){\n    return (ms) ? _defer(\"timeout\", callback, ms) : _defer(\"immediate\", callback)\n}\n\ndefer.immediate = function(callback){\n    return _defer(\"immediate\", callback)\n}\n\ndefer.timeout = function(callback, ms){\n    if (!ms) ms = 0\n    return _defer(\"timeout\", callback, ms)\n}\n\ndefer.frame = function(callback){\n    return _defer(\"frame\", callback)\n}\n\nmodule.exports = defer\n");
__$coverInitRange("defer.js", "11:24");
__$coverInitRange("defer.js", "25:94");
__$coverInitRange("defer.js", "95:107");
__$coverInitRange("defer.js", "108:480");
__$coverInitRange("defer.js", "481:710");
__$coverInitRange("defer.js", "711:773");
__$coverInitRange("defer.js", "774:793");
__$coverInitRange("defer.js", "794:1454");
__$coverInitRange("defer.js", "1457:1743");
__$coverInitRange("defer.js", "1744:1853");
__$coverInitRange("defer.js", "1854:2060");
__$coverInitRange("defer.js", "2061:2181");
__$coverInitRange("defer.js", "2182:2263");
__$coverInitRange("defer.js", "2264:2369");
__$coverInitRange("defer.js", "2370:2443");
__$coverInitRange("defer.js", "2444:2466");
__$coverInitRange("defer.js", "157:231");
__$coverInitRange("defer.js", "232:252");
__$coverInitRange("defer.js", "253:322");
__$coverInitRange("defer.js", "328:362");
__$coverInitRange("defer.js", "363:389");
__$coverInitRange("defer.js", "390:420");
__$coverInitRange("defer.js", "421:477");
__$coverInitRange("defer.js", "288:301");
__$coverInitRange("defer.js", "311:316");
__$coverInitRange("defer.js", "448:474");
__$coverInitRange("defer.js", "521:557");
__$coverInitRange("defer.js", "558:575");
__$coverInitRange("defer.js", "576:668");
__$coverInitRange("defer.js", "674:707");
__$coverInitRange("defer.js", "611:642");
__$coverInitRange("defer.js", "643:667");
__$coverInitRange("defer.js", "743:770");
__$coverInitRange("defer.js", "839:914");
__$coverInitRange("defer.js", "881:912");
__$coverInitRange("defer.js", "952:1023");
__$coverInitRange("defer.js", "994:1021");
__$coverInitRange("defer.js", "1088:1288");
__$coverInitRange("defer.js", "1289:1367");
__$coverInitRange("defer.js", "1141:1269");
__$coverInitRange("defer.js", "1213:1248");
__$coverInitRange("defer.js", "1249:1268");
__$coverInitRange("defer.js", "1331:1364");
__$coverInitRange("defer.js", "1381:1453");
__$coverInitRange("defer.js", "1423:1451");
__$coverInitRange("defer.js", "1706:1740");
__$coverInitRange("defer.js", "1778:1850");
__$coverInitRange("defer.js", "1820:1847");
__$coverInitRange("defer.js", "1892:1926");
__$coverInitRange("defer.js", "1927:1981");
__$coverInitRange("defer.js", "1982:2042");
__$coverInitRange("defer.js", "2012:2035");
__$coverInitRange("defer.js", "2101:2178");
__$coverInitRange("defer.js", "2224:2260");
__$coverInitRange("defer.js", "2308:2327");
__$coverInitRange("defer.js", "2328:2366");
__$coverInitRange("defer.js", "2408:2440");
__$coverCall('defer.js', '11:24');
'use strict';
__$coverCall('defer.js', '25:94');
var deferred = {
        timeout: {},
        frame: {},
        immediate: {}
    };
__$coverCall('defer.js', '95:107');
var UID = 0;
__$coverCall('defer.js', '108:480');
var _defer = function (type, callback, arg) {
    __$coverCall('defer.js', '157:231');
    var collection = deferred[type], method = functions[type];
    __$coverCall('defer.js', '232:252');
    var empty = true;
    __$coverCall('defer.js', '253:322');
    for (var p in collection) {
        __$coverCall('defer.js', '288:301');
        empty = false;
        __$coverCall('defer.js', '311:316');
        break;
    }
    __$coverCall('defer.js', '328:362');
    var uid = (UID++).toString(36);
    __$coverCall('defer.js', '363:389');
    if (empty)
        method(arg);
    __$coverCall('defer.js', '390:420');
    collection[uid] = callback;
    __$coverCall('defer.js', '421:477');
    return function () {
        __$coverCall('defer.js', '448:474');
        delete collection[uid];
    };
};
__$coverCall('defer.js', '481:710');
var iterate = function (collection) {
    __$coverCall('defer.js', '521:557');
    var time = new Date().getTime();
    __$coverCall('defer.js', '558:575');
    var exec = {};
    __$coverCall('defer.js', '576:668');
    for (var p in collection) {
        __$coverCall('defer.js', '611:642');
        exec[p] = collection[p];
        __$coverCall('defer.js', '643:667');
        delete collection[p];
    }
    __$coverCall('defer.js', '674:707');
    for (var p in exec)
        exec[p](time);
};
__$coverCall('defer.js', '711:773');
var immediate = function () {
    __$coverCall('defer.js', '743:770');
    iterate(deferred.immediate);
};
__$coverCall('defer.js', '774:793');
var functions = {};
__$coverCall('defer.js', '794:1454');
if (global.process && process.nextTick) {
    __$coverCall('defer.js', '839:914');
    functions.immediate = function () {
        __$coverCall('defer.js', '881:912');
        process.nextTick(immediate);
    };
} else if (global.setImmediate) {
    __$coverCall('defer.js', '952:1023');
    functions.immediate = function () {
        __$coverCall('defer.js', '994:1021');
        setImmediate(immediate);
    };
} else if (global.postMessage && global.addEventListener) {
    __$coverCall('defer.js', '1088:1288');
    addEventListener('message', function (event) {
        __$coverCall('defer.js', '1141:1269');
        if (event.source === global && event.data === '@deferred') {
            __$coverCall('defer.js', '1213:1248');
            event.stopPropagation();
            __$coverCall('defer.js', '1249:1268');
            immediate();
        }
    }, true);
    __$coverCall('defer.js', '1289:1367');
    functions.immediate = function () {
        __$coverCall('defer.js', '1331:1364');
        postMessage('@deferred', '*');
    };
} else {
    __$coverCall('defer.js', '1381:1453');
    functions.immediate = function () {
        __$coverCall('defer.js', '1423:1451');
        setTimeout(immediate, 0);
    };
}
__$coverCall('defer.js', '1457:1743');
var requestAnimationFrame = global.requestAnimationFrame || global.webkitRequestAnimationFrame || global.mozRequestAnimationFrame || global.oRequestAnimationFrame || global.msRequestAnimationFrame || function (callback) {
        __$coverCall('defer.js', '1706:1740');
        setTimeout(callback, 1000 / 60);
    };
__$coverCall('defer.js', '1744:1853');
functions.frame = function () {
    __$coverCall('defer.js', '1778:1850');
    requestAnimationFrame(function () {
        __$coverCall('defer.js', '1820:1847');
        iterate(deferred.frame);
    });
};
__$coverCall('defer.js', '1854:2060');
functions.timeout = function (ms) {
    __$coverCall('defer.js', '1892:1926');
    var timeout = deferred.timeout;
    __$coverCall('defer.js', '1927:1981');
    var collection = timeout[ms] || (timeout[ms] = {});
    __$coverCall('defer.js', '1982:2042');
    setTimeout(function () {
        __$coverCall('defer.js', '2012:2035');
        iterate(collection);
    }, ms);
};
__$coverCall('defer.js', '2061:2181');
var defer = function (callback, ms) {
    __$coverCall('defer.js', '2101:2178');
    return ms ? _defer('timeout', callback, ms) : _defer('immediate', callback);
};
__$coverCall('defer.js', '2182:2263');
defer.immediate = function (callback) {
    __$coverCall('defer.js', '2224:2260');
    return _defer('immediate', callback);
};
__$coverCall('defer.js', '2264:2369');
defer.timeout = function (callback, ms) {
    __$coverCall('defer.js', '2308:2327');
    if (!ms)
        ms = 0;
    __$coverCall('defer.js', '2328:2366');
    return _defer('timeout', callback, ms);
};
__$coverCall('defer.js', '2370:2443');
defer.frame = function (callback) {
    __$coverCall('defer.js', '2408:2440');
    return _defer('frame', callback);
};
__$coverCall('defer.js', '2444:2466');
module.exports = defer;
