// prime/cov/object.js
var __$coverCall, __$coverInit, __$coverInitRange;
(function(fs){
	"use strict";
	var update = function(cb){
		return function(a, b, c, d){
			var json = JSON.parse(fs.readFileSync("/home/arian/www/prime/cov.json").toString() || '{}');
			cb(json, a, b);
			fs.writeFileSync("/home/arian/www/prime/cov.json", JSON.stringify(json, null, 2));
		};
	};
	__$coverInit = update(function(json, name, code){
		if (!json[name]) json[name] = {__code: code};
	});
	__$coverInitRange = update(function(json, name, range){
		if (!json[name][range]) json[name][range] = 0;
	});
	__$coverCall = update(function(json, name, range){
		json[name][range]++;
	});
})(require('fs'));
__$coverInit("object.js", "/*\nobject\n*/\"use strict\"\n\nvar object = require(\"./_shell\")[\"object\"]\nvar proto = Object.prototype\n\nobject.implement({\n    hasOwnProperty: proto.hasOwnProperty,\n    toString:       proto.toString,\n    valueOf:        proto.valueOf\n})\n\nobject.extend({\n    count:   require(\"./object/count\"),\n    create:  require(\"./object/create\"),\n    every:   require(\"./object/every\"),\n    filter:  require(\"./object/filter\"),\n    forIn:   require(\"./object/forIn\"),\n    forOwn:  require(\"./object/forOwn\"),\n    hasOwn:  require(\"./object/hasOwn\"),\n    indexOf: require(\"./object/indexOf\"),\n    keys:    require(\"./object/keys\"),\n    map:     require(\"./object/map\"),\n    mixIn:   require(\"./object/mixIn\"),\n    remove:  require(\"./object/remove\"),\n    some:    require(\"./object/some\"),\n    unset:   require(\"./object/unset\"),\n    values:  require(\"./object/values\")\n})\n\nmodule.exports = object\n");
__$coverInitRange("object.js", "12:25");
__$coverInitRange("object.js", "26:68");
__$coverInitRange("object.js", "69:98");
__$coverInitRange("object.js", "99:233");
__$coverInitRange("object.js", "234:856");
__$coverInitRange("object.js", "857:880");
__$coverCall('object.js', '12:25');
'use strict';
__$coverCall('object.js', '26:68');
var object = require('./_shell')['object'];
__$coverCall('object.js', '69:98');
var proto = Object.prototype;
__$coverCall('object.js', '99:233');
object.implement({
    hasOwnProperty: proto.hasOwnProperty,
    toString: proto.toString,
    valueOf: proto.valueOf
});
__$coverCall('object.js', '234:856');
object.extend({
    count: require('./object/count'),
    create: require('./object/create'),
    every: require('./object/every'),
    filter: require('./object/filter'),
    forIn: require('./object/forIn'),
    forOwn: require('./object/forOwn'),
    hasOwn: require('./object/hasOwn'),
    indexOf: require('./object/indexOf'),
    keys: require('./object/keys'),
    map: require('./object/map'),
    mixIn: require('./object/mixIn'),
    remove: require('./object/remove'),
    some: require('./object/some'),
    unset: require('./object/unset'),
    values: require('./object/values')
});
__$coverCall('object.js', '857:880');
module.exports = object;
