// prime/cov/string.js
var __$coverCall, __$coverInit, __$coverInitRange;
(function(fs){
	"use strict";
	var update = function(cb){
		return function(a, b, c, d){
			var json = JSON.parse(fs.readFileSync("/home/arian/www/prime/cov.json").toString() || '{}');
			cb(json, a, b);
			fs.writeFileSync("/home/arian/www/prime/cov.json", JSON.stringify(json, null, 2));
		};
	};
	__$coverInit = update(function(json, name, code){
		if (!json[name]) json[name] = {__code: code};
	});
	__$coverInitRange = update(function(json, name, range){
		if (!json[name][range]) json[name][range] = 0;
	});
	__$coverCall = update(function(json, name, range){
		json[name][range]++;
	});
})(require('fs'));
__$coverInit("string.js", "/*\nstring\n*/\"use strict\"\n\nvar string = require(\"./_shell\")[\"string\"]\nvar proto = String.prototype\n\nstring.implement({\n    charAt: proto.charAt,\n    charCodeAt: proto.charCodeAt,\n    concat: proto.concat,\n    indexOf: proto.indexOf,\n    lastIndexOf: proto.lastIndexOf,\n    match: proto.match,\n    replace: proto.replace,\n    search: proto.search,\n    slice: proto.slice,\n    split: proto.split,\n    substr: proto.substr,\n    substring: proto.substring,\n    toLowerCase: proto.toLowerCase,\n    toUpperCase: proto.toUpperCase,\n    toString: proto.toString,\n    valueOf: proto.valueOf\n})\n\nstring.extend({\n    camelize: require(\"./string/camelize\"),\n    capitalize: require(\"./string/capitalize\"),\n    clean: require(\"./string/clean\"),\n    escape: require(\"./string/escape\"),\n    hyphenate: require(\"./string/hyphenate\"),\n    trim: require(\"./string/trim\")\n})\n\nmodule.exports = string\n");
__$coverInitRange("string.js", "12:25");
__$coverInitRange("string.js", "26:68");
__$coverInitRange("string.js", "69:98");
__$coverInitRange("string.js", "99:584");
__$coverInitRange("string.js", "585:855");
__$coverInitRange("string.js", "856:879");
__$coverCall('string.js', '12:25');
'use strict';
__$coverCall('string.js', '26:68');
var string = require('./_shell')['string'];
__$coverCall('string.js', '69:98');
var proto = String.prototype;
__$coverCall('string.js', '99:584');
string.implement({
    charAt: proto.charAt,
    charCodeAt: proto.charCodeAt,
    concat: proto.concat,
    indexOf: proto.indexOf,
    lastIndexOf: proto.lastIndexOf,
    match: proto.match,
    replace: proto.replace,
    search: proto.search,
    slice: proto.slice,
    split: proto.split,
    substr: proto.substr,
    substring: proto.substring,
    toLowerCase: proto.toLowerCase,
    toUpperCase: proto.toUpperCase,
    toString: proto.toString,
    valueOf: proto.valueOf
});
__$coverCall('string.js', '585:855');
string.extend({
    camelize: require('./string/camelize'),
    capitalize: require('./string/capitalize'),
    clean: require('./string/clean'),
    escape: require('./string/escape'),
    hyphenate: require('./string/hyphenate'),
    trim: require('./string/trim')
});
__$coverCall('string.js', '856:879');
module.exports = string;
